File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/CsrSegmentMaxBp.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Backward pass for {@link CsrSegmentMax}: gradient of the neighbourhood max-aggregation.
 *
 * <p>C++ op name: {@code csr_segment_max_bp}
 * <ul>
 *   <li>Inputs:  colIdx[nnz] (INT32), rowPtr[rows+1] (INT32),
 *                X[n, f] (forward input node features),
 *                gradOut[rows, f] (upstream gradient)</li>
 *   <li>IArgs:   rows</li>
 *   <li>Output:  dX[n, f] — gradient w.r.t. the node feature matrix X</li>
 * </ul>
 *
 * <p>Gradient semantics: for each row {@code i} and feature {@code f}, the gradient
 * {@code gradOut[i, f]} is routed back to the unique neighbour {@code j*} that achieved
 * the maximum {@code X[j*, f]} in the forward pass.  If a node wins the max in multiple
 * rows, its gradient accumulates (sum).  Ties are broken by the native op.
 *
 * <p>This is a backward primitive: {@code doDiff} is not implemented.
 */
public class CsrSegmentMaxBp extends DynamicCustomOp {

    /** No-arg constructor required for op-registry reflection. */
    public CsrSegmentMaxBp() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param colIdx   1D [nnz] INT32 column (source-node) indices
     * @param rowPtr   1D [rows+1] INT32 row (segment) pointers
     * @param X        2D [n, f] dense node feature matrix from the forward pass
     * @param gradOut  2D [rows, f] upstream gradient
     * @param rows     number of rows (segments)
     */
    public CsrSegmentMaxBp(INDArray colIdx, INDArray rowPtr, INDArray X,
                           INDArray gradOut, long rows) {
        super(new INDArray[]{colIdx, rowPtr, X, gradOut}, null);
        addIArgument(rows);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd       the SameDiff graph
     * @param colIdx   SDVariable [nnz] INT32 column indices
     * @param rowPtr   SDVariable [rows+1] INT32 row pointers
     * @param X        SDVariable [n, f] forward node features
     * @param gradOut  SDVariable [rows, f] upstream gradient
     * @param rows     number of rows
     */
    public CsrSegmentMaxBp(SameDiff sd, SDVariable colIdx, SDVariable rowPtr,
                           SDVariable X, SDVariable gradOut, long rows) {
        super(sd, new SDVariable[]{colIdx, rowPtr, X, gradOut});
        addIArgument(rows);
    }

    @Override
    public String opName() { return "csr_segment_max_bp"; }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        // dX: same dtype as X (input[2])
        return Collections.singletonList(dataTypes.get(2));
    }
}

File: libnd4j/include/ops/declarable/generic/sparse/csr_segment_max_bp.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// csr_segment_max_bp — backward pass of csr_segment_max.
//
// For each (row i, feature feat):
//   Recompute k* = argmax_{k in row i} X[colIdx[k], feat].
//   dX[colIdx[k*], feat] += gradOut[i, feat].
//
// Multiple rows can map to the same (node, feat) → atomicAdd in CUDA.
// dX is initialised to zero (OUTPUT_NULLIFIED); CUDA helper also calls
// cudaMemsetAsync to guarantee the device buffer is zero before scatter.
//
// Inputs:
//   [0] colIdx   [nnz]     int    — CSR column indices
//   [1] rowPtr   [rows+1]  int    — CSR row pointer array
//   [2] X        [n, f]    float  — original node features
//   [3] gradOut  [rows, f] float  — upstream gradient
// IArgs:
//   [0] rows
// Outputs:
//   [0] dX  [n, f]  float  — gradient w.r.t. node features
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_csr_segment_max_bp)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_segment_max.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(csr_segment_max_bp, 4, 1, false, 0, 1) {
  auto colIdx  = INPUT_VARIABLE(0);  // [nnz]     int
  auto rowPtr  = INPUT_VARIABLE(1);  // [rows+1]  int
  auto X       = INPUT_VARIABLE(2);  // [n, f]    float
  auto gradOut = INPUT_VARIABLE(3);  // [rows, f] float

  const sd::LongType rows = INT_ARG(0);
  const sd::LongType f    = X->sizeAt(1);

  REQUIRE_TRUE(colIdx->rankOf() == 1, 0,
               "csr_segment_max_bp: colIdx must be rank-1, got %d", colIdx->rankOf());
  REQUIRE_TRUE(rowPtr->rankOf() == 1, 0,
               "csr_segment_max_bp: rowPtr must be rank-1, got %d", rowPtr->rankOf());
  REQUIRE_TRUE(X->rankOf() == 2, 0,
               "csr_segment_max_bp: X must be rank-2 [n, f], got %d", X->rankOf());
  REQUIRE_TRUE(gradOut->rankOf() == 2, 0,
               "csr_segment_max_bp: gradOut must be rank-2 [rows, f], got %d",
               gradOut->rankOf());
  REQUIRE_TRUE(gradOut->sizeAt(0) == rows && gradOut->sizeAt(1) == f, 0,
               "csr_segment_max_bp: gradOut must be [%lld, %lld], got [%lld, %lld]",
               (long long)rows, (long long)f,
               (long long)gradOut->sizeAt(0), (long long)gradOut->sizeAt(1));
  REQUIRE_TRUE(rowPtr->lengthOf() == rows + 1, 0,
               "csr_segment_max_bp: rowPtr length must be rows+1=%lld, got %lld",
               (long long)(rows + 1), (long long)rowPtr->lengthOf());

  auto dX = OUTPUT_NULLIFIED(0);  // [n, f] — zeroed by nullification + helper

  sd::ops::helpers::csr_segment_max_bp(*colIdx, *rowPtr, *X, *gradOut, *dX, rows);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(csr_segment_max_bp) {
  // dX has the same shape and dtype as X
  auto X = INPUT_VARIABLE(2);
  auto dxShape = ConstantShapeHelper::getInstance().createShapeInfo(
      X->dataType(), 'c', {X->sizeAt(0), X->sizeAt(1)});
  return SHAPELIST(dxShape);
}

DECLARE_TYPES(csr_segment_max_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_INTS})    // colIdx
      ->setAllowedInputTypes(1, {ALL_INTS})    // rowPtr
      ->setAllowedInputTypes(2, {ALL_FLOATS})  // X
      ->setAllowedInputTypes(3, {ALL_FLOATS})  // gradOut
      ->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_csr_segment_max_bp)
